src/Build/Evaluation/Expander.cs

```diff
@@ -4677,6 +4694,40 @@ private static bool TryGetArgs(object[] args, out string arg0, out int arg1)
                 return false;
             }
 
+            private static bool IsFloatingPointRepresentation(object value)
+            {
+                const char numberDecimalSeparator = '.';
+
+                return value is double || (value is string str && str.Contains(numberDecimalSeparator));
+            }
+
+            private static bool TryExecuteArithmeticOverload(object[] args, Func<long, long, long> integerOperation, Func<double, double, double> realOperation, out object resultValue)
+            {
+                resultValue = null;
+
+                if (args.Length != 2)
+                {
+                    return false;
+                }
+
+                if (IsIntrinsicFunctionOverloadsEnabled() && !IsFloatingPointRepresentation(args[0]) && !IsFloatingPointRepresentation(args[1]))
+                {
+                    if (TryConvertToLong(args[0], out long argLong0) && TryConvertToLong(args[1], out long argLong1))
+                    {
+                        resultValue = integerOperation(argLong0, argLong1);
```

Comment:
Your comment that even without the fast path, we were still only doing double arithmetic made me think hard about problems like this. Specifically, what is the "right" behavior when we would overflow post-operation? So like what is the correct value for Long.MaxValue + 1?

I'm wondering whether it's a better design to prioritize long arithmetic, as we suggest we do in the docs, and say Long.MaxValue + 1 = Long.MinValue or quietly convert it to a double before doing the operation. Sorry if that was already discussed, and I missed it.

Reply:
I suspect that converting to `double` on overflow would be adding a feature that may never be used in the field. I expect there is very little arithmetic performed in MSBuild and that most or nearly all of the arithmetic is with integer values that are well within the min and max values of `long`.

Reply:
Since there is a workaround for this edge case (explicitly define any operand as double), I'm fine with the behavior. It just should bee properly documented